### PR TITLE
cmake: use cmake includedir

### PR DESCRIPTION
**Pull request by tripzero**

### CODE REVIEW

**Comment by tripzero** [comment]
something still isn't quite right... plz wait.

**Comment by arfoll** [comment]
Happy to wait but can we get more details in the commit message as to why this change is needed/wanted and what problem it solves?

**Comment by tripzero** [comment]
Sure.  Will do.

**Comment by tripzero** [comment]
The problem space is more complex than this patch fixes.  The issue is within yocto and cmake itself.  Within cmake, CMAKE_INSTALL_PREFIX has a unique meaning[1].  It means the same as DESTDIR when running "make DESTDIR=/foo install".  In yocto, this variable is being inappropriately set to ${prefix} which is not the same as DESTDIR (yocto probably uses this elsewhere).  The result is that many files are being installed inappropriately: /usr/etc instead of /etc, /usr/lib/systemd instead of /lib/systemd, etc.

If we fix cmake.bbclass in yocto, a bunch of packages that expect ${CMAKE_INSTALL_PREFIX} including mraa.

In mraa, we can fix this by using GnuInstallDirs or the default cmake install variables instead of hardcoding with the expectation that CMAKE_INSTALL_PREFIX is /usr.

The fact that cmake makes "/usr/local" the default prefix is disturbing to me.  I think it gives developers a false assumption about its meaning.

[1] - https://cmake.org/cmake/help/v3.2/variable/CMAKE_INSTALL_PREFIX.html

**Comment by tripzero** [comment]
I changed the commit message to explain the rationale.  There's one more patch I've got coming but I will make a separate commit because I don't really like the patch and I think more discussion is required to "Do It Right" TM.

**Comment by tripzero** [comment]
@arfoll I removed the nodejs change.  Any comments on the python changes?

**Comment by alext-mkrs** [comment]
Actually I still see a patch with nodejs changes in this pull request, are you sure you've removed it?

**Comment by tripzero** [comment]
@alext-mkrs check the commit again.  I don't see any nodejs changes.

**Comment by alext-mkrs** [comment]
And now I don't see that commit either  so it's ok. I think it would also help if you describe the algorithm you want to use here.

**Comment by tripzero** [comment]
The algorithm for the python site-package path finding?  My algorithm uses the getsitepackages() python call to get the OS site package path and then replaces the prefix.

The problem with the old method was that it insalled to "site-packages" on Debian (and debian-like distros) and Debian doesn't use site-packages so python can't find mraa.

File: src/python/CMakeLists.txt [hunk 1]
```diff
@@ -38,12 +38,33 @@ execute_process (
...
 execute_process (
    COMMAND ${PYTHON_EXECUTABLE} -c
-       "import site, sys; sys.stdout.write(site.getusersitepackages().replace(site.getuserbase(), site.PREFIXES[-1]))"
+       "	\n
+import site, sys \n
+dirs = site.getsitepackages(); \n
+dir = dirs[0] \n
+for d in dirs: \n
+	if d.count(\"packages\"): \n
+		dir = d[d.index(\"packages\")-5:] \n
```
Comment (alext-mkrs) [inline src/python/CMakeLists.txt]: this "-5" will break if the directory for modules is called e.g. "usersite-packages", i.e. when the part before "packages" is not five characters long.

File: src/python/CMakeLists.txt [hunk 2]
```diff
@@ -38,12 +38,33 @@ execute_process (
...
+		break \n
+py = \"${PYTHON_EXECUTABLE}\" \n
+py = py[py.rfind(\"/\")+1:] \n
+prefix = \"${PYTHON_PREFIX}\"[\"${PYTHON_PREFIX}\".rfind(\"/\"):] \n
+dir = \"{0}/lib/{1}/{2}\".format(prefix, py, dir) \n
```
Comment (alext-mkrs) [inline src/python/CMakeLists.txt]: `py` will not work on e.g. Edison - where the executable is `/usr/bin/python2`, but the packages dir is `/usr/lib/python2.7/site-packages`, i.e. Python executable name is not equal to `/usr/lib` subfolder.

File: src/python/CMakeLists.txt [hunk 3]
```diff
@@ -38,12 +38,33 @@ execute_process (
...
 )
-file (TO_CMAKE_PATH "${PYTHON_SITE_DIR}" PYTHON_SITE_DIR)
-string (REGEX REPLACE "^${PYTHON_PREFIX}/" ""
-   PYTHON_SITE_DIR "${PYTHON_SITE_DIR}")
+#"import site, sys; sys.stdout.write(site.getsitepackages()[0])"
+
+message(STATUS "python site dirs: ${PYTHON_SITE_DIR}")
+message(STATUS "python prefix: ${PYTHON_PREFIX}")
+
+if("${CMAKE_INSTALL_PREFIX}" STREQUAL "/usr")
```
Comment (alext-mkrs) [inline src/python/CMakeLists.txt]: What if it's not `/usr`?

Could you please clarify the point of this whole block? The `PYTHON_SITE_DIR` variable is being set to `lib/python2.7/dist-packages` on my Debian 8 after this is executed, so the `install()` function will try to install it into current directory.

**Comment by alext-mkrs** [comment]
I've just checked your patch out and while I think the correctiong in `examples/CMakeLists.txt` and `src/CMakeLists.txt` make sense, I would say the Python part might need changes, as it IMHO replaces current inoptimal approach to another one, which happens to work for current Debian, but has some drawbacks I've tried to highlight in per-line comments.

We definitely need to fix that Python install part (it was already a problem mentioned in #336, just in some more optimal way. I'll see if I can come up with some proposal as well.

**Comment by tripzero** [comment]
Alext, fair enough.  I don't like the python changes and your criticisms are valid.  I will remove that part of the patch and resubmit.

**Comment by arfoll** [comment]
Merged, excuse the delay!

**Pull request closed without merging**